src/runtime.cc

```diff
@@ -105,233 +105,323 @@ namespace detail {
+				CELERITY_INFO("Task graph:\n\n{}\n", tdag_str);
 			}
-			// must be called on all nodes
-			auto cmd_graph = gather_command_graph();
+
+			assert(m_command_recorder.get() != nullptr);
+			auto cdag_str = print_command_graph(m_local_nid, *m_command_recorder);
+			if(!is_dry_run()) { cdag_str = gather_command_graph(cdag_str, m_num_nodes, m_local_nid); } // must be called on all nodes
+
 			if(m_local_nid == 0) {
-				std::this_thread::sleep_for(std::chrono::milliseconds(500)); // Avoid racing on stdout with other nodes (funneled through mpirun)
-				CELERITY_INFO("Command graph:\n\n{}\n", cmd_graph);
+				// Avoid racing on stdout with other nodes (funneled through mpirun)
+				if(!is_dry_run()) { std::this_thread::sleep_for(std::chrono::milliseconds(500)); }
+				CELERITY_INFO("Command graph:\n\n{}\n", cdag_str);
+			}
+
+			// IDAGs become unreadable when all nodes print them at the same time - TODO attempt gathering them as well?
+			if(m_local_nid == 0) {
+				// we are allowed to deref m_instruction_recorder / m_command_recorder because the scheduler thread has exited at this point
+				const auto idag_str = detail::print_instruction_graph(*m_instruction_recorder, *m_command_recorder, *m_task_recorder);
+				CELERITY_INFO("Instruction graph on node 0:\n\n{}\n", idag_str);
 			}
 		}
 
-		// Shutting down the task_manager will cause all buffers captured inside command group functions to unregister.
-		// Since we check whether the runtime is still active upon unregistering, we have to set this to false first.
-		m_is_active = false;
-		m_task_mngr->shutdown();
-		m_is_shutting_down = false;
-		maybe_destroy_runtime();
+		m_instruction_recorder.reset();
+		m_command_recorder.reset();
+		m_task_recorder.reset();
+
+		cgf_diagnostics::teardown();
+
+		m_user_bench.reset();
+
+		if(!s_test_mode) { mpi_finalize_once(); }
 	}
 
-	void runtime::sync() {
-		const auto epoch = m_task_mngr->generate_epoch_task(epoch_action::barrier);
+	void runtime::sync(epoch_action action) {
+		require_call_from_application_thread();
+		const auto epoch = m_task_mngr->generate_epoch_task(action);
 		m_task_mngr->await_epoch(epoch);
 	}
 
-	task_manager& runtime::get_task_manager() const { return *m_task_mngr; }
-
-	buffer_manager& runtime::get_buffer_manager() const { return *m_buffer_mngr; }
-
-	reduction_manager& runtime::get_reduction_manager() const { return *m_reduction_mngr; }
+	task_manager& runtime::get_task_manager() const {
+		require_call_from_application_thread();
+		return *m_task_mngr;
+	}
 
-	std::string runtime::gather_command_graph() const {
-		assert(m_command_recorder.get() != nullptr);
-		const auto graph_str = print_command_graph(m_local_nid, *m_command_recorder);
+	std::string gather_command_graph(const std::string& graph_str, const size_t num_nodes, const node_id local_nid) {
+		const auto comm = MPI_COMM_WORLD;
+		const int tag = 'CDAG'; // Celerity does not perform any other peer-to-peer communication over MPI_COMM_WORLD
```

Comment:
This is cute, but I locally get a `warning: multi-character character constant [-Wmultichar]` for it.

Reply:
Outsmarted again. I've increased the level of l33t.